src/haddock/libs/libgrid.py

```diff
@@ -395,119 +393,401 @@ def __repr__(self) -> str:
+            self.payload_fnames.append(inp_name)
+            with open(inp_name, "w") as f:
+                for line in recipe_str.splitlines(keepends=True):
+                    # Parse this line and try to identify output files
+                    output = self._find_output(line)
+                    if output:
+                        self.expected_outputs.append(output)
+
+                    # Process the line to adjust paths, if any
+                    new_line, found_fname = self._process_line(line)
+
+                    # Collect the files that need to be in the payload
+                    if found_fname:
+                        src_path = Path(found_fname)
+                        dst_path = self.loc / Path(found_fname).name
+                        shutil.copy(src_path, dst_path)
+                        self.payload_fnames.append(dst_path)
+
+                    f.write(new_line)
+
+
 class GRIDScheduler:
     """Scheduler to manage and run jobs on the GRID via DIRAC."""
 
-    tasks: list[CNSJob]
-    params: dict
+    def __init__(
+        self, tasks: list[CNSJob], params: dict, probing: float = 0.05
+    ) -> None:
+
+        self.probing: bool = True
+        self.ncores = parse_ncores(params["ncores"])
+        self.batch_size = 1
+
+        self.workload: list[GridJob] = [
+            GridJob(
+                input=t.input_file,
+                toppar_path=t.envvars["TOPPAR"],
+                module_path=t.envvars["MODULE"],
+            )
+            for t in tasks
+        ]
+
+        # ===============================================================#
+        #
+        # ! IMPORTANT !
+        #
+        # The `subset_size` is the number of jobs to be used for probing the grid
+        # In theory we could send any number of jobs, but in practice we need to
+        #  consider that the submission itself takes time. So we can come to the
+        #  scenario in which we spend more time submitting jobs than measuring,
+        #  this means that we would not be able to properly evaluate the grid
+        #  capacity.
+        # Here we set the `subset_size` to be the minimum between `ncores`
+        # and the 5% of the total number of jobs. By setting it to the number
+        # of cores, we can be sure that we will be able to measure the metrics.
+        # We also need to ensure there is at least one job to probe the grid.
+        #
+        # HACK: This is not ideal, but it is a compromise between accuracy and speed.
+        subset_size = min(self.ncores, max(1, math.ceil(len(self.workload) * probing)))
+        # ===============================================================#
+
+        # Randomly select that many jobs
+        if len(self.workload) >= subset_size:
+            subset_keys = random.sample(list(self.workload), subset_size)
+            for job in self.workload:
+                if job in subset_keys:
+                    job.tag = Tag.PROBING
```

Comment:
Usually it is better to ask for indices when using random.sample (especially true when the list is very long, which may not be the case in haddock).
This also allows then to skip the condition after, as we loop only on the selected indices
```suggestion
            subset_keys = random.sample(list(self.workload), subset_size)
            for job in self.workload:
                if job in subset_keys:
                    job.tag = Tag.PROBING
```
```suggestion
            subset_indices = random.sample(range(len(self.workload)), subset_size)
            for i in subset_indices:
                self.workload[i].tag = Tag.PROBING
```